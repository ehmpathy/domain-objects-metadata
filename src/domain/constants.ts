import { createIsOfEnum } from 'type-fns';

export enum DomainObjectVariant {
  DOMAIN_OBJECT = 'DomainObject',
  DOMAIN_LITERAL = 'DomainLiteral',
  DOMAIN_ENTITY = 'DomainEntity',
  DOMAIN_EVENT = 'DomainEvent',
}
export const isOfDomainObjectVariant = createIsOfEnum(DomainObjectVariant);
