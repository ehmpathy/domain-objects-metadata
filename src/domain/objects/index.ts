export * from './DomainObjectMetadata';
export * from './DomainObjectMetadataReference';
export * from './DomainObjectProperty';
