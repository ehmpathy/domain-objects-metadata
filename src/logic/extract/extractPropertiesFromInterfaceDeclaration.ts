import { InterfaceDeclaration, isArrayTypeNode, isTypeReferenceNode, Node, SyntaxKind, TypeElement } from 'typescript';
import { DomainObjectProperty, DomainObjectPropertyType } from '../../domain/objects/DomainObjectMetadata';

interface ASTInterfacePropertyType extends Node {
  name: { escapedText: string };
  kind: SyntaxKind;
  types?: ASTInterfacePropertyType[]; // present on unions
  elementType?: ASTInterfacePropertyType; // present on arrays
}

const extractPrimaryTypeFromMemberDeclaration = ({
  memberDeclaration,
  propertyName,
  interfaceName,
}: {
  memberDeclaration: TypeElement;
  propertyName: string;
  interfaceName: string;
}): { primaryType: ASTInterfacePropertyType; nullable: boolean; required: boolean } => {
  // grab the first level type on the membership declaration
  const rootType: ASTInterfacePropertyType = (memberDeclaration as any).type;

  // figure out whether its required
  const required = !memberDeclaration.questionToken;

  // if its not a union, then its not nullable and the firstLevelType is the primary type
  if (rootType.kind !== SyntaxKind.UnionType) return { primaryType: rootType, required, nullable: false };

  // if it is a union, then look at the subtypes.
  const subTypes = rootType.types;
  if (!subTypes) throw new Error(`root type is a UnionType but does not have subtypes. this is unexpected. see ${interfaceName}.${propertyName}`);

  // it should only have two, and one of them will be the `NullKeyword` type
  const hasMoreThanTwoSubtypes = subTypes.length > 2;
  const oneOfTheSubtypesIsNotNull = !subTypes.some((type) => type.kind === SyntaxKind.NullKeyword);
  if (hasMoreThanTwoSubtypes || oneOfTheSubtypesIsNotNull)
    throw new Error(
      `domain object property types can only have one primary type. they can be nullable or optional, but they can not be 'string | number', for example. not satisfied by ${interfaceName}.${propertyName} `,
    );
  const typeOtherThanNull = subTypes.find((type) => type.kind !== SyntaxKind.NullKeyword)!;
  return { primaryType: typeOtherThanNull, nullable: true, required };
};

const extractPropertyDefinitionFromNormalizedMemberTypeDefinition = ({
  primaryType,
  nullable,
  required,
  propertyName,
  interfaceName,
}: {
  primaryType: ASTInterfacePropertyType;
  nullable: boolean;
  required: boolean;
  propertyName: string;
  interfaceName: string;
}): DomainObjectProperty => {
  // handle the simple cases first
  if (primaryType.kind === SyntaxKind.StringKeyword) return new DomainObjectProperty({ type: DomainObjectPropertyType.STRING, nullable, required });
  if (primaryType.kind === SyntaxKind.NumberKeyword) return new DomainObjectProperty({ type: DomainObjectPropertyType.NUMBER, nullable, required });
  if (primaryType.kind === SyntaxKind.BooleanKeyword) return new DomainObjectProperty({ type: DomainObjectPropertyType.BOOLEAN, nullable, required });

  // handle the array case
  if (isArrayTypeNode(primaryType))
    return new DomainObjectProperty({
      type: DomainObjectPropertyType.ARRAY,
      of: extractPropertyDefinitionFromNormalizedMemberTypeDefinition({
        primaryType: primaryType.elementType,
        nullable: false, // elements of an array should not be null, nulls should be filtered out
        required: true, // elements of an array should not be undefined, undefineds should be filtered out
        propertyName,
        interfaceName,
      }),
      nullable,
      required,
    });

  // handle the nested type reference
  if (isTypeReferenceNode(primaryType))
    return new DomainObjectProperty({
      type: DomainObjectPropertyType.REFERENCE,
      of: (primaryType.typeName as any).escapedText, // note: this is a string
      nullable,
      required,
    });

  // throw an error otherwise
  throw new Error(`could not extract property definition from interface member declaration. see ${interfaceName}.${propertyName}`);
};

const extractPropertyFromDomainObjectInterfaceMemberDeclaration = ({
  memberDeclaration,
  interfaceName,
}: {
  memberDeclaration: TypeElement;
  interfaceName: string;
}): { name: string; definition: DomainObjectProperty } => {
  // grab the name
  const propertyName = (memberDeclaration.name as any).escapedText;

  // normalize the type data on the member declaration
  const { primaryType, nullable, required } = extractPrimaryTypeFromMemberDeclaration({ memberDeclaration, propertyName, interfaceName });

  // grab the property based on the normalized type data now
  const definition = extractPropertyDefinitionFromNormalizedMemberTypeDefinition({ primaryType, nullable, required, interfaceName, propertyName });

  return {
    name: propertyName,
    definition,
  };
};

export const extractPropertiesFromInterfaceDeclaration = (interfaceDeclaration: InterfaceDeclaration) => {
  const properties = interfaceDeclaration.members.map((memberDeclaration) =>
    extractPropertyFromDomainObjectInterfaceMemberDeclaration({
      memberDeclaration,
      interfaceName: interfaceDeclaration.name.text,
    }),
  );
  const propertiesObject: { [index: string]: DomainObjectProperty } = {};
  properties.forEach((property) => {
    propertiesObject[property.name] = property.definition;
  });
  return propertiesObject;
};
