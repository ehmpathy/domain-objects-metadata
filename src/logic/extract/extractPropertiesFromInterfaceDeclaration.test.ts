import ts, { isInterfaceDeclaration } from 'typescript';
import { DomainObjectPropertyType } from '../../domain';
import { extractPropertiesFromInterfaceDeclaration } from './extractPropertiesFromInterfaceDeclaration';

describe('extractPropertiesFromInterfaceDeclaration', () => {
  it('should be able to extract properties from our Address interface. all strings', () => {
    const program = ts.createProgram([`${__dirname}/../__test_assets__/Address.ts`], {});
    const file = program.getSourceFiles().find((thisFile) => thisFile.fileName.includes('/Address.ts'))!; // grab the right file
    const interfaceDeclaration = file.statements.find(isInterfaceDeclaration)!;
    const properties = extractPropertiesFromInterfaceDeclaration(interfaceDeclaration);
    // console.log(JSON.stringify(properties, null, 2));
    expect(properties).toMatchSnapshot();
  });
  it('should be able to extract properties from an interface with all of the primitive types we support: number, string, boolean w/ required and optional combination', () => {
    const program = ts.createProgram([`${__dirname}/../__test_assets__/LoadingDock.ts`], {});
    const file = program.getSourceFiles().find((thisFile) => thisFile.fileName.includes('/LoadingDock.ts'))!; // grab the right file
    const interfaceDeclaration = file.statements.find(isInterfaceDeclaration)!;
    const properties = extractPropertiesFromInterfaceDeclaration(interfaceDeclaration);
    // console.log(JSON.stringify(properties, null, 2));

    // prove we got every combo
    expect(Object.values(properties).map((property) => property.type)).toContain(DomainObjectPropertyType.STRING);
    expect(Object.values(properties).map((property) => property.type)).toContain(DomainObjectPropertyType.NUMBER);
    expect(Object.values(properties).map((property) => property.type)).toContain(DomainObjectPropertyType.BOOLEAN);
    expect(Object.values(properties).map((property) => property.required)).toContain(true);
    expect(Object.values(properties).map((property) => property.required)).toContain(false);
    expect(Object.values(properties).map((property) => property.nullable)).toContain(true);
    expect(Object.values(properties).map((property) => property.nullable)).toContain(false);

    // save an example
    expect(properties).toMatchSnapshot();
  });
  it('should be able to extract properties from an interface with an array property', () => {
    const program = ts.createProgram([`${__dirname}/../__test_assets__/Item.ts`], {});
    const file = program.getSourceFiles().find((thisFile) => thisFile.fileName.includes('/Item.ts'))!; // grab the right file
    const interfaceDeclaration = file.statements.find(isInterfaceDeclaration)!;
    const properties = extractPropertiesFromInterfaceDeclaration(interfaceDeclaration);
    // console.log(JSON.stringify(properties, null, 2));

    // prove we got the array defined correctly
    const arrayProperty = Object.values(properties).find((property) => property.type === DomainObjectPropertyType.ARRAY);
    if (!arrayProperty) throw new Error('array property not defined');
    expect(arrayProperty.of).toEqual({ type: DomainObjectPropertyType.STRING, nullable: false, required: true });

    // save an example
    expect(properties).toMatchSnapshot();
  });
  it('should be able to extract properties from an interface that references another interface or an array of interfaces', () => {
    const program = ts.createProgram([`${__dirname}/../__test_assets__/Order.ts`], {});
    const file = program.getSourceFiles().find((thisFile) => thisFile.fileName.includes('/Order.ts'))!; // grab the right file
    const interfaceDeclaration = file.statements.find(isInterfaceDeclaration)!;
    const properties = extractPropertiesFromInterfaceDeclaration(interfaceDeclaration);
    // console.log(JSON.stringify(properties, null, 2));

    // prove we got the nested interfaces defined referenced correctly
    expect(properties.destination).toMatchObject({
      type: DomainObjectPropertyType.REFERENCE,
      of: 'Address',
    });
    expect(properties.items).toMatchObject({
      type: DomainObjectPropertyType.ARRAY,
      of: {
        type: DomainObjectPropertyType.REFERENCE,
        of: 'Item',
      },
    });

    // save an example
    expect(properties).toMatchSnapshot();
  });
  it('should be able to extract properties from an interface with an enum', () => {
    const program = ts.createProgram([`${__dirname}/../__test_assets__/Payment.ts`], {});
    const file = program.getSourceFiles().find((thisFile) => thisFile.fileName.includes('/Payment.ts'))!; // grab the right file
    const interfaceDeclaration = file.statements.find(isInterfaceDeclaration)!;
    const properties = extractPropertiesFromInterfaceDeclaration(interfaceDeclaration);
    // console.log(JSON.stringify(properties, null, 2));

    // prove we got the enum defined
    expect(properties.status).toMatchObject({
      type: DomainObjectPropertyType.REFERENCE, // notice how because it is a type reference, the most we can tell at this point is that its a reference. we'll have to hydrate to cast into enum
      of: 'PaymentStatus',
    });

    // save an example
    expect(properties).toMatchSnapshot();
  });
  it('should be able to extract properties from an interface with a date type', () => {
    const program = ts.createProgram([`${__dirname}/../__test_assets__/DeliveredEvent.ts`], {});
    const file = program.getSourceFiles().find((thisFile) => thisFile.fileName.includes('/DeliveredEvent.ts'))!; // grab the right file
    const interfaceDeclaration = file.statements.find(isInterfaceDeclaration)!;
    const properties = extractPropertiesFromInterfaceDeclaration(interfaceDeclaration);
    // console.log(JSON.stringify(properties, null, 2));

    // prove we got the enum defined
    expect(properties.occurredAt).toMatchObject({ type: DomainObjectPropertyType.DATE });

    // save an example
    expect(properties).toMatchSnapshot();
  });
});
