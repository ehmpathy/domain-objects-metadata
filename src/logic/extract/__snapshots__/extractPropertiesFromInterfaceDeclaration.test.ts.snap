// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extractPropertiesFromInterfaceDeclaration should be able to extract properties from an interface that references another interface or an array of interfaces 1`] = `
Object {
  "destination": DomainObjectPropertyMetadata {
    "name": "destination",
    "nullable": false,
    "of": "Address",
    "required": true,
    "type": "REFERENCE",
  },
  "id": DomainObjectPropertyMetadata {
    "name": "id",
    "nullable": false,
    "required": false,
    "type": "NUMBER",
  },
  "items": DomainObjectPropertyMetadata {
    "name": "items",
    "nullable": false,
    "of": Object {
      "nullable": false,
      "of": "Item",
      "required": true,
      "type": "REFERENCE",
    },
    "required": true,
    "type": "ARRAY",
  },
  "status": DomainObjectPropertyMetadata {
    "name": "status",
    "nullable": false,
    "required": true,
    "type": "STRING",
  },
  "uuid": DomainObjectPropertyMetadata {
    "name": "uuid",
    "nullable": false,
    "required": false,
    "type": "STRING",
  },
}
`;

exports[`extractPropertiesFromInterfaceDeclaration should be able to extract properties from an interface with a date type 1`] = `
Object {
  "id": DomainObjectPropertyMetadata {
    "name": "id",
    "nullable": false,
    "required": false,
    "type": "NUMBER",
  },
  "occurredAt": DomainObjectPropertyMetadata {
    "name": "occurredAt",
    "nullable": false,
    "required": true,
    "type": "DATE",
  },
  "packageUuid": DomainObjectPropertyMetadata {
    "name": "packageUuid",
    "nullable": false,
    "required": true,
    "type": "STRING",
  },
  "uuid": DomainObjectPropertyMetadata {
    "name": "uuid",
    "nullable": false,
    "required": false,
    "type": "STRING",
  },
}
`;

exports[`extractPropertiesFromInterfaceDeclaration should be able to extract properties from an interface with all of the primitive types we support: number, string, boolean w/ required and optional combination 1`] = `
Object {
  "gateNumber": DomainObjectPropertyMetadata {
    "name": "gateNumber",
    "nullable": false,
    "required": true,
    "type": "NUMBER",
  },
  "id": DomainObjectPropertyMetadata {
    "name": "id",
    "nullable": false,
    "required": false,
    "type": "NUMBER",
  },
  "operational": DomainObjectPropertyMetadata {
    "name": "operational",
    "nullable": false,
    "required": true,
    "type": "BOOLEAN",
  },
  "phoneNumber": DomainObjectPropertyMetadata {
    "name": "phoneNumber",
    "nullable": true,
    "required": true,
    "type": "STRING",
  },
  "uuid": DomainObjectPropertyMetadata {
    "name": "uuid",
    "nullable": false,
    "required": false,
    "type": "STRING",
  },
}
`;

exports[`extractPropertiesFromInterfaceDeclaration should be able to extract properties from an interface with an array property 1`] = `
Object {
  "colors": DomainObjectPropertyMetadata {
    "name": "colors",
    "nullable": false,
    "of": Object {
      "nullable": false,
      "required": true,
      "type": "STRING",
    },
    "required": true,
    "type": "ARRAY",
  },
  "description": DomainObjectPropertyMetadata {
    "name": "description",
    "nullable": false,
    "required": true,
    "type": "STRING",
  },
  "price": DomainObjectPropertyMetadata {
    "name": "price",
    "nullable": false,
    "required": true,
    "type": "NUMBER",
  },
  "upc": DomainObjectPropertyMetadata {
    "name": "upc",
    "nullable": false,
    "required": true,
    "type": "STRING",
  },
}
`;

exports[`extractPropertiesFromInterfaceDeclaration should be able to extract properties from an interface with an enum 1`] = `
Object {
  "amount": DomainObjectPropertyMetadata {
    "name": "amount",
    "nullable": false,
    "required": true,
    "type": "NUMBER",
  },
  "currency": DomainObjectPropertyMetadata {
    "name": "currency",
    "nullable": false,
    "required": true,
    "type": "STRING",
  },
  "externalId": DomainObjectPropertyMetadata {
    "name": "externalId",
    "nullable": false,
    "required": true,
    "type": "STRING",
  },
  "status": DomainObjectPropertyMetadata {
    "name": "status",
    "nullable": false,
    "of": "PaymentStatus",
    "required": true,
    "type": "REFERENCE",
  },
}
`;

exports[`extractPropertiesFromInterfaceDeclaration should be able to extract properties from our Address interface. all strings 1`] = `
Object {
  "city": DomainObjectPropertyMetadata {
    "name": "city",
    "nullable": false,
    "required": true,
    "type": "STRING",
  },
  "postal": DomainObjectPropertyMetadata {
    "name": "postal",
    "nullable": false,
    "required": true,
    "type": "STRING",
  },
  "state": DomainObjectPropertyMetadata {
    "name": "state",
    "nullable": false,
    "required": true,
    "type": "STRING",
  },
  "street": DomainObjectPropertyMetadata {
    "name": "street",
    "nullable": false,
    "required": true,
    "type": "STRING",
  },
  "suite": DomainObjectPropertyMetadata {
    "name": "suite",
    "nullable": true,
    "required": true,
    "type": "STRING",
  },
}
`;
